docs/AQL/grammar/AqlParser.g4

```diff
@@ -0,0 +1,189 @@
+//  contributors: This version of the grammar is a complete rewrite of previously published antlr3 grammar,
+//                based on current AQL specifications in combination with other grammars from several AQL implementations.
+//                The openEHR Foundation would like to recognise the following people for their contributions:
+//                  - Chunlan Ma & Heath Frankel, Ocen Health Systems, Australia
+//                  - Bostjan Lah, Better, Slovenia
+//                  - Christian Chevalley, EHRBase, Germany
+//                  - Michael Böckers, Nedap, Netherlands
+//  support:      openEHR Specifications PR tracker <https://specifications.openehr.org/releases/QUERY/open_issues>
+//  copyright:    Copyright (c) 2020 openEHR Foundation
+//  license:      Apache 2.0 License <http://www.apache.org/licenses/LICENSE-2.0.html>
+//
+
+parser grammar AqlParser;
+
+options { tokenVocab=AqlLexer; }
+
+selectQuery
+    : selectClause fromClause whereClause? orderByClause? limitClause? SYM_DOUBLE_DASH? EOF
+    ;
+
+selectClause
+    : SELECT top? selectExpr (SYM_COMMA selectExpr)*
+    ;
+
+fromClause
+    : FROM fromExpr
+    ;
+
+whereClause
+    : WHERE whereExpr
+    ;
+
+orderByClause
+    : ORDER BY orderByExpr (SYM_COMMA orderByExpr)*
+    ;
+
+limitClause
+    : LIMIT limit=NATURAL_NUMBER (OFFSET offset=WHOLE_NUMBER) ?
+    ;
+
+
+selectExpr
+    : columnExpr (AS aliasName=ALIAS_ID)?
+    ;
+
+fromExpr
+    : containsExpr
+    ;
+
+whereExpr
+    : NOT? identifiedExpr
+    | whereExpr AND whereExpr
+    | whereExpr OR whereExpr
+    | SYM_LEFT_PAREN whereExpr SYM_RIGHT_PAREN
+    ;
+
+orderByExpr
+    : identifiedPath order=(DESCENDING|DESC|ASCENDING|ASC)?
+    ;
+
+
+columnExpr
+    : identifiedPath
+    | primitive
+    | aggregateFunctionCall
+    | functionCall
+    ;
+
+containsExpr
+    : classExprOperand (NOT? CONTAINS containsExpr)?
+    | containsExpr AND containsExpr
+    | containsExpr OR containsExpr
+    | SYM_LEFT_PAREN containsExpr SYM_RIGHT_PAREN
+    ;
+
+identifiedExpr
+    : EXISTS identifiedPath
+    | identifiedPath COMPARISON_OPERATOR terminal
+    | functionCall COMPARISON_OPERATOR terminal
+    | identifiedPath LIKE likeOperand
+    | identifiedPath MATCHES SYM_LEFT_CURLY matchesOperand SYM_RIGHT_CURLY
+    ;
+
+classExprOperand
+    : TYPE_ID variable=VARIABLE_ID? (SYM_LEFT_BRACKET archetypePredicate SYM_RIGHT_BRACKET)?                    #classExpression // RM_TYPE_NAME variable [archetype_id]
+    | VERSIONED_OBJECT variable=VARIABLE_ID? (SYM_LEFT_BRACKET standardPredicate SYM_RIGHT_BRACKET)?            #versionedClassExpr
+    | EHR variable=VARIABLE_ID? (SYM_LEFT_BRACKET standardPredicate SYM_RIGHT_BRACKET)?                         #ehrClassExpr
+    | VERSION variable=VARIABLE_ID? (SYM_LEFT_BRACKET (standardPredicate|versionPredicate) SYM_RIGHT_BRACKET)?  #versionClassExpr
+    ;
+
+terminal
+    : primitive
+    | PARAMETER
+    | identifiedPath
+    | functionCall
+    ;
+
+identifiedPath
```

Comment:
I think it would still be clearer to call this identifiedObject, because its value is not a path, it's an object, which is also what the other 'terminal' types have as their values. An xxxPath is something I would expect to be able to concatente with another path, for example.

Reply:
In fact I am not sure if this is correct anyway; because there is a difference between where you declare an object id with a predicate and a path, and where you use it. I think the correct rules might be more like:
```
objectRelPath:
     pathPart (SYM_SLASH pathPart)*
    ;

objectPath:
      SYM_SLASH objectPath
    ;

identifedObjectDecl:
    VARIABLE_ID SYM_LEFT_BRACKET archetypePredicate SYM_RIGHT_BRACKET  objectPath
    ;

identifiedObjectRef:
    VARIABLE_ID
    ;
```

I need to study this more to get it correct however.

Reply:
just a remark here: I tried to use as much as possible names that are present in specifications, as I thought it might be easier for the reader/programmer to match rules and things derived from this rules or token to written textual concepts in specifications.

Reply:
> I think it would still be clearer to call this identifiedObject, because its value is not a path, it's an object, which is also what the other 'terminal' types have as their values. An xxxPath is something I would expect to be able to concatente with another path, for example.

what I intended to capture there with the rule is an [identified path](https://specifications.openehr.org/releases/QUERY/latest/AQL.html#_identified_paths_syntax) - see examples:
- o/data[at0001]/.../data[at0003]/items[at0004]/value/value
- o[name/value=$nameValue]
- o[name/value=$nameValue]/data[at0001]/.../data[at0003]/items[at0004]/value/value

Reply:
If you don't mind, I will like to keep it this way, until there is a going to be teste and clear report that this is incorrect or more feedback